### PR TITLE
[Backport 2.x][Search Pipelines] Support ad hoc pipelines #7253

### PR DESCRIPTION
**Pull request by msfroh**
### Description
Backport https://github.com/opensearch-project/OpenSearch/pull/7253 to 2.x (since auto-backport failed).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by reta — APPROVED** [review]

File: server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -340,6 +351,12 @@ public void writeTo(StreamOutput out) throws IOException {
         if (out.getVersion().onOrAfter(LegacyESVersion.V_7_10_0)) {
             out.writeOptionalWriteable(pointInTimeBuilder);
         }
+        if (out.getVersion().onOrAfter(Version.V_2_8_0)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:354]: As always, we would need to bring it back to `main` :)

Comment (msfroh) [reply server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:354]: Yup! 

I might piggy-back that into https://github.com/opensearch-project/OpenSearch/pull/7470

**Comment by kotwanikunal** [comment]
@reta / @msfroh Can you please mark the conversations as resolved before we can merge it in?

**Comment by msfroh** [comment]
> Can you please mark the conversations as resolved before we can merge it in?

@kotwanikunal -- done!

**Pull request merged**